server/identification.ts

```diff
@@ -66,31 +66,56 @@ class Identification {
 
 	serverConnection(socket: Socket) {
 		socket.on("error", (err: string) => log.error(`Identd socket error: ${err}`));
-		socket.on("data", (data) => {
+		socket.setTimeout(5000, () => {
+			log.warn(
+				`identd: no data received, closing connection to ${
+					socket.remoteAddress || "undefined"
+				}`
+			);
+			socket.destroy();
+		});
+		socket.once("data", (data) => {
 			this.respondToIdent(socket, data);
 			socket.end();
 		});
 	}
 
 	respondToIdent(socket: Socket, buffer: Buffer) {
+		if (!socket.remoteAddress) {
+			log.warn("identd: no remote address");
+			return;
+		}
+
 		const data = buffer.toString().split(",");
 
 		const lport = parseInt(data[0], 10) || 0;
 		const fport = parseInt(data[1], 10) || 0;
 
 		if (lport < 1 || fport < 1 || lport > 65535 || fport > 65535) {
+			log.warn(`identd: bogus request from ${socket.remoteAddress}`);
 			return;
 		}
 
+		log.debug(`identd: remote ${socket.remoteAddress} query ${lport}, ${fport}`);
+
 		for (const connection of this.connections.values()) {
-			if (connection.socket.remotePort === fport && connection.socket.localPort === lport) {
-				return socket.write(
-					`${lport}, ${fport} : USERID : TheLounge : ${connection.user}\r\n`
-				);
+			// we only want to respond if all the ip,port tuples match, to avoid user enumeration
+			if (
+				connection.socket.remotePort === fport &&
+				connection.socket.localPort === lport &&
+				socket.remoteAddress === connection.socket.remoteAddress &&
+				socket.localAddress === connection.socket.localAddress
```

Comment:
This ignores ipv4 in ipv6 handling...
Meaning if for some reason our ident server listens on ipv6 (dual stack) but the IRC connection is in ipv4 we will wrongly send an not found response.

Considering that you have to try rather hard to get to that edge case, I chose not to care.